File: player.py
# player.py
import math
import pygame
from settings import *


class Player:
    def __init__(self, game):
        self.game = game
        self.x = TILE_SIZE * 2
        self.y = TILE_SIZE * 2
        self.angle = -math.pi / 2

        if not pygame.joystick.get_init():
            pygame.joystick.init()

        self.joystick = None
        self.start_btn_pressed = False

        self.shoot_cooldown = 0
        self.shockwave_cooldown = 0

        self.connect_joystick()

    def connect_joystick(self):
        if pygame.joystick.get_count() > 0:
            try:
                self.joystick = pygame.joystick.Joystick(0)
                self.joystick.init()
                print(f"Joystick connecté : {self.joystick.get_name()}")
            except Exception as e:
                print(f"Erreur connexion joystick : {e}")
                self.joystick = None
        else:
            print("Aucun joystick détecté pour le moment.")
            self.joystick = None

    def disconnect_joystick(self):
        if self.joystick:
            print(f"Déconnexion du joystick : {self.joystick.get_name()}")
            self.joystick.quit()
            self.joystick = None

    def is_action_pressed(self, action, keys, mouse_btns, joystick):

        b_type, b_val = self.game.parsed_controls['Bindings'].get(action, (None, None))
        if b_type == 'key' and b_val != -1 and keys[b_val]:
            return True
        if b_type == 'mouse' and b_val != -1:
            idx = b_val - 1
            if 0 <= idx < len(mouse_btns) and mouse_btns[idx]:
                return True

        b_type_pad, b_val_pad = self.game.parsed_controls['Gamepad'].get(action, (None, None))
        if b_type_pad == 'btn' and b_val_pad != -1 and joystick and joystick.get_init():
            if joystick.get_numbuttons() > b_val_pad and joystick.get_button(b_val_pad):
                return True

        return False

    def movement(self):
        dx = 0
        dy = 0
        current_speed = PLAYER_SPEED

        keys = pygame.key.get_pressed()
        mouse_btns = pygame.mouse.get_pressed()

        if self.shoot_cooldown > 0: self.shoot_cooldown -= 1
        if self.shockwave_cooldown > 0: self.shockwave_cooldown -= 1

        if self.is_action_pressed('sprint', keys, mouse_btns, self.joystick):
            current_speed *= 2.0
        if self.is_action_pressed('slow', keys, mouse_btns, self.joystick):
            current_speed *= 0.5

        if self.is_action_pressed('forward', keys, mouse_btns, self.joystick): dy -= current_speed
        if self.is_action_pressed('backward', keys, mouse_btns, self.joystick): dy += current_speed
        if self.is_action_pressed('left', keys, mouse_btns, self.joystick): dx -= current_speed
        if self.is_action_pressed('right', keys, mouse_btns, self.joystick): dx += current_speed

        # Rotation
        if self.is_action_pressed('rot_left', keys, mouse_btns, self.joystick): self.angle -= PLAYER_ROT_SPEED
        if self.is_action_pressed('rot_right', keys, mouse_btns, self.joystick): self.angle += PLAYER_ROT_SPEED

        # Actions
        if self.is_action_pressed('shoot', keys, mouse_btns, self.joystick): self.shoot()
        if self.is_action_pressed('shockwave', keys, mouse_btns, self.joystick): self.cast_shockwave()

        if self.joystick and self.joystick.get_init():
            try:
                ax0 = self.joystick.get_axis(0)
                ax1 = self.joystick.get_axis(1)

                deadzone = 0.15

                if abs(ax0) > deadzone:
                    dx += ax0 * current_speed
                if abs(ax1) > deadzone:
                    dy += ax1 * current_speed

                rs_x, rs_y = 0, 0
                if self.joystick.get_numaxes() >= 4:
                    rs_x = self.joystick.get_axis(2)
                    rs_y = self.joystick.get_axis(3)

                if math.sqrt(rs_x ** 2 + rs_y ** 2) > 0.5:
                    target_angle = math.atan2(rs_y, rs_x)
                    self.angle = target_angle

            except pygame.error:
                self.disconnect_joystick()

        self.check_collision(dx, dy)

    def shoot(self):
        if self.shoot_cooldown <= 0:
            self.game.entity_manager.add_projectile(self.x, self.y, self.angle)
            self.shoot_cooldown = 15

    def cast_shockwave(self):
        if self.shockwave_cooldown <= 0:
            self.game.entity_manager.apply_shockwave(self.x, self.y)
            self.shockwave_cooldown = 60

    def check_collision(self, dx, dy):
        scale = PLAYER_SIZE * 0.5

        if dx != 0:
            new_x = self.x + dx
            if self.game.map_handler.is_position_free(new_x, self.y, scale):
                self.x += dx

        if dy != 0:
            new_y = self.y + dy
            if self.game.map_handler.is_position_free(self.x, new_y, scale):
                self.y += dy

    @property
    def pos(self):
        return (self.x, self.y)

    @property
    def map_pos(self):
        return (int(self.x // TILE_SIZE), int(self.y // TILE_SIZE))

File: src/player.rs
// src/player.rs
use crate::config::*;
use crate::map::MapHandler;
use crate::input::InputState;

pub struct Player {
    pub x: f32,
    pub y: f32,
    pub angle: f32,
    pub shoot_cooldown: i32,
    pub shockwave_cooldown: i32,
}

impl Player {
    pub fn new() -> Self {
        Self {
            x: TILE_SIZE * 2.0,
            y: TILE_SIZE * 2.0,
            angle: -std::f32::consts::FRAC_PI_2,
            shoot_cooldown: 0,
            shockwave_cooldown: 0,
        }
    }

    pub fn pos(&self) -> (f32, f32) {
        (self.x, self.y)
    }

    pub fn map_pos(&self) -> (i32, i32) {
        ((self.x / TILE_SIZE) as i32, (self.y / TILE_SIZE) as i32)
    }

    pub fn update(&mut self, input: &InputState, map: &mut MapHandler, dt: f32) -> (bool, bool) {
        let mut dx = 0.0f32;
        let mut dy = 0.0f32;
        // Vitesse en pixels/seconde, normalisée à 60fps pour conserver le feeling
        let mut current_speed = PLAYER_SPEED * dt * 60.0;
        let mut wants_shoot = false;
        let mut wants_shockwave = false;

        if self.shoot_cooldown > 0 { self.shoot_cooldown -= 1; }
        if self.shockwave_cooldown > 0 { self.shockwave_cooldown -= 1; }

        if input.sprint { current_speed *= 2.0; }
        if input.slow { current_speed *= 0.5; }

        if input.forward { dy -= current_speed; }
        if input.backward { dy += current_speed; }
        if input.left { dx -= current_speed; }
        if input.right { dx += current_speed; }

        if input.rot_left { self.angle -= PLAYER_ROT_SPEED * dt * 60.0; }
        if input.rot_right { self.angle += PLAYER_ROT_SPEED * dt * 60.0; }

        // Click gauche : orienter le joueur vers la souris et avancer dans cette direction
        if input.mouse_left_down {
            if let Some(aim) = input.aim_angle {
                self.angle = aim;
                dx += aim.cos() * current_speed;
                dy += aim.sin() * current_speed;
            }
        }

        if input.shoot {
            wants_shoot = self.try_shoot();
        }
        if input.shockwave {
            wants_shockwave = self.try_shockwave();
        }

        // Gamepad analog sticks
        if input.move_x.abs() > 0.15 {
            dx += input.move_x * current_speed;
        }
        if input.move_y.abs() > 0.15 {
            dy += input.move_y * current_speed;
        }

        // Right stick aim
        let rs_mag = (input.look_x * input.look_x + input.look_y * input.look_y).sqrt();
        if rs_mag > 0.5 {
            self.angle = input.look_y.atan2(input.look_x);
        }

        // Collision check
        self.check_collision(dx, dy, map);

        (wants_shoot, wants_shockwave)
    }

    fn try_shoot(&mut self) -> bool {
        if self.shoot_cooldown <= 0 {
            self.shoot_cooldown = 45;
            true
        } else {
            false
        }
    }

    fn try_shockwave(&mut self) -> bool {
        if self.shockwave_cooldown <= 0 {
            self.shockwave_cooldown = 60;
            true
        } else {
            false
        }
    }

    fn check_collision(&mut self, dx: f32, dy: f32, map: &mut MapHandler) {
        let scale = PLAYER_SIZE * 0.5;

        if dx != 0.0 {
            let new_x = self.x + dx;
            if map.is_position_free(new_x, self.y, scale) {
                self.x = new_x;
            }
        }

        if dy != 0.0 {
            let new_y = self.y + dy;
            if map.is_position_free(self.x, new_y, scale) {
                self.y = new_y;
            }
        }
    }
}
